### PR TITLE
Quadratic module fix

### PR DESCRIPTION
**Pull request by tweisser**
Fix the issue that the variables of q had to be contained in the variables of p when creating the quadratic module. 

Example: (old version)
p = x + 1, q = 1 - x^2 - y^2
It does make sense to ask whether p is positive on the set q>=0, even though p does not depend on y.
However, when I do
```julia
using SumOfSquares
using DynamicPolynomials
using CSDP 

@polyvar x
@polyvar y
m = SOSModel(with_optimizer(CSDP.Optimizer))
@variable m t
@objective m Max t
@constraint m sosc x - t in SOSCone() domain = @set(1 - x^2 - y^2 >=0)
optimize!(m)
```
I get ERROR: AssertionError: variables(q) ⊆ variables(p)

But clearly t =  -1 is (optimal and) feasible  as
x - ( - 1 ) = 1/2*( x + 1 )^2 + 1/2y^2  + 1/2*( 1 - x^2 - y^2 )

(this version)
After the fix optimize!(m) works and  CSDP solves the problem:
```julia
objective_value(m)
# -1.0000000003752996
```
The Mock tests fail, because the generated SDP is different now.

### CODE REVIEW

File: src/sos_polynomial_in_semialgebraic_set_bridge.jl [hunk 1]
```diff
@@ -9,10 +9,9 @@ function lagrangian_multiplier(model::MOI.ModelLike, p, set::SOSLikeCone, q,
     # If maxdegree_s2 is odd, div(maxdegree_s2,2) would make s^2 have degree up to maxdegree_s2-1
     # for this reason, we take div(maxdegree_s2+1,2) so that s^2 have degree up to maxdegree_s2+1
     maxdegree_s = div(maxdegree_s2 + 1, 2)
-    # FIXME handle the case where `p`, `q_i`, ...  do not have the same variables
-    # so instead of `variable(p)` we would have the union of them all
-    @assert variables(q) ⊆ variables(p)
-    monos = MP.monomials(MP.variables(p), mindegree_s:maxdegree_s)
+    vars = [MP.variables(p)..., MP.variables(q)...]
+    unique!(vars)
+    monos = MP.monomials(sort(vars, rev=true), 0:maxdegree_s)
```
Comment (blegat) [inline src/sos_polynomial_in_semialgebraic_set_bridge.jl]: Why use `0:maxdegree_s` instead of `mindegree_s:maxdegree_s` ?

Comment (blegat) [reply src/sos_polynomial_in_semialgebraic_set_bridge.jl]: This is probably what makes the mock tests fail

Comment (tweisser) [reply src/sos_polynomial_in_semialgebraic_set_bridge.jl]: Maybe I did not understand what mindegree in the Input is. I assumed it to be the minimal degree of p.  But this doesn't make sense. Say p is homogeneous of degree 4 and q is quadratic. Then mindegree(p) = maxdegree(p) = 4 and mindegree(q)=0, maxdegree(q)=2. Consequently, mindegree_s2 = 4 and maxdegree_s2 = 2, which is definititly wrong...

File: src/sos_polynomial_in_semialgebraic_set_bridge.jl [hunk 2]
```diff
@@ -9,10 +9,9 @@ function lagrangian_multiplier(model::MOI.ModelLike, p, set::SOSLikeCone, q,
     # If maxdegree_s2 is odd, div(maxdegree_s2,2) would make s^2 have degree up to maxdegree_s2-1
     # for this reason, we take div(maxdegree_s2+1,2) so that s^2 have degree up to maxdegree_s2+1
     maxdegree_s = div(maxdegree_s2 + 1, 2)
-    # FIXME handle the case where `p`, `q_i`, ...  do not have the same variables
-    # so instead of `variable(p)` we would have the union of them all
-    @assert variables(q) ⊆ variables(p)
-    monos = MP.monomials(MP.variables(p), mindegree_s:maxdegree_s)
+    vars = [MP.variables(p)..., MP.variables(q)...]
+    unique!(vars)
```
Comment (blegat) [inline src/sos_polynomial_in_semialgebraic_set_bridge.jl]: According to the docstring of `unique!`, it is more efficient to do `sort!` then `unique!`:
```julia
julia> unique!([1, 1, 1])
  1-element Array{Int64,1}:
   1
  
  julia> A = [7, 3, 2, 3, 7, 5];
  
  julia> unique!(A)
  4-element Array{Int64,1}:
   7
   3
   2
   5
  
  julia> B = [7, 6, 42, 6, 7, 42];
  
  julia> sort!(B);  # unique! is able to process sorted data much more efficiently.
  
  julia> unique!(B)
  3-element Array{Int64,1}:
    6
    7
   42
```

File: src/sos_polynomial_in_semialgebraic_set_bridge.jl (hunk 1 above)
Comment (blegat) [reply src/sos_polynomial_in_semialgebraic_set_bridge.jl]: Indeed, we should probably use 0 here do you have a good example to add as a test ?

Comment (tweisser) [reply src/sos_polynomial_in_semialgebraic_set_bridge.jl]: Here is an example that shows that 0:maxdegree_s is necessary:

    @polyvar x y
    m = SOSModel(factory)
    p = x^4 - y^4 
    S = @set - x^2 - y^4 >=0
    soscon = @constraint m p >= 0 domain = S 
    JuMP.optimize!(m)

One possible sos_decomposition is
(x)^2 + (x^2)^2  + (1)^2 * (-y^4 - x^2)
However, mindegree_s is 1 in this example and indeed, on master the SDP is infeasible.

File: src/sos_polynomial_in_semialgebraic_set_bridge.jl [hunk 3]
```diff
@@ -9,10 +9,9 @@ function lagrangian_multiplier(model::MOI.ModelLike, p, set::SOSLikeCone, q,
     # If maxdegree_s2 is odd, div(maxdegree_s2,2) would make s^2 have degree up to maxdegree_s2-1
     # for this reason, we take div(maxdegree_s2+1,2) so that s^2 have degree up to maxdegree_s2+1
     maxdegree_s = div(maxdegree_s2 + 1, 2)
-    # FIXME handle the case where `p`, `q_i`, ...  do not have the same variables
-    # so instead of `variable(p)` we would have the union of them all
-    @assert variables(q) ⊆ variables(p)
-    monos = MP.monomials(MP.variables(p), mindegree_s:maxdegree_s)
+    vars = sort([MP.variables(p)..., MP.variables(q)...])
```
Comment (blegat) [inline src/sos_polynomial_in_semialgebraic_set_bridge.jl]: `sort!` is more efficient as it will do it in-place instead of creating a new array

**Pull request merged**